Feature/header

kangminhyuk1111 opened the pull request:
header, sidebar 1차 작업 완료에 대한 pr입니다.

추후에 컴포넌트에 인자로 받는 빈 배열에 리스트들을 넣어주면 됩니다.
ListType은 title과 route로 이루어져 있습니다.

/test/route/1 여기 경로로 들어가시면 테스트 페이지 구현되어있습니다.

다음 dropdown 컴포넌트 작업하고 있겠습니다.

kangminhyuk1111 committed:
개인push

kangminhyuk1111 committed:
Merge branch 'zockq1:main' into feature/header

heegu0311 committed:
config: git ignore

kangminhyuk1111 committed:
Merge pull request #27 from zockq1/config/gitignore

config: git ignore

kangminhyuk1111 committed:
header component화 작업중
로컬 push

kangminhyuk1111 committed:
header component화 작업중
로컬 push

kangminhyuk1111 committed:
header component화 작업중
로컬 push

kangminhyuk1111 committed:
FEAT: Header,Sidebar 1차 작업 완료

zockq1 commented:
1. 각 컴포넌트 이름에 Component 제거하는게 나아보입니다. (예를 들어 SidebarComponent -> Sidebar)
2. 각 컴포넌트에 useEffect에 console.log 찍는거 머지하기 전에는 제거 부탁드립니다.
3. 지금 ui/header/Header.tsx, layout/Header.tsx 두개 있는데 하나로 합치는게 좋아보입니다. 
4. 그리고 Header 관련된건 ui디렉토리 말고 layout 디렉토리에 header 디렉토리 만들어서 옮겨주시면 좋겠습니다.
5. unknown 타입들도 머지 전에는 전부 interface로 정의 부탁드립니다

heegu0311 committed:
Merge pull request #14 from zockq1/feature/breadcrumb

Feat: #13 Breadcrumb 추가

heegu0311 commented:
@kangminhyuk1111 

중복 코드를 줄일 수 있도록 컴포넌트화 하는건 어떨까요?

```tsx
// Header.tsx 의 아래 부분과

<li>
  <Link className="hover:text-blue-500 text-sm" href="/">
    홈
  </Link>
</li>

// HeaderDetail.tsx 의 아래 부분

<li key={index}>
  <Link href={str.route}>{str.title}</Link>
</li>
```

```tsx
// HeaderDetail.tsx 내 6종 대분류 Nav 메뉴

<li className="bottom-1 mb-1 border-solid border-black text-center text-lg font-bold">
  홈
</li>

<li className="bottom-1 mb-1 border-solid border-black text-center text-lg font-bold">
  콘텐츠
</li>

```

heegu0311 commented on src/share/ui/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,44 @@
+}: {
+  sidebarList: unknown[];
+}) {
+  const router = usePathname();
```
@kangminhyuk1111 아래 Next에서 제공해주는 다른 훅을 사용하는건 어떨까요?
아시겠지만 App router 부분에 폴더 네이밍 [testId] 이런 부분의 정보를 쉽게 가져올수 있게 해줍니다~!

usePathname -> [useParams](https://nextjs.org/docs/app/api-reference/functions/use-params)

heegu0311 commented on src/share/ui/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,44 @@
+}) {
+  const router = usePathname();
+  const activeSegments = router.split('/')[3];
+  const linkQuery: string[] = ['1', '2', '3', '4', '5'];
```
linkQuery를 별도로 선언해서 사용하기보다는 contentList 내에 `route` 속성값과 매칭시키는 방식으로 처리하면 코드가 간결해지지 않을까요?

<img width="1523" alt="image" src="https://github.com/zockq1/piggymy-frontend-admin/assets/13115713/f3d81717-c6e8-46e6-8e61-802b0e494ae8">

heegu0311 commented:
마지막으로 [zockq1:develop](https://github.com/zockq1/piggymy-frontend-admin/tree/develop) 기준으로 싱크를 맞춘 후에 추가 작업을 하시는 건 어떨지요~~! forked 레포라 그게 가능한지는 모르겠지만..

현재 develop 브랜치에 실제 사용하게 될 route 정보가 포함되어 있기 때문에 싱크 후에 작업하시면 편리할 것 같습니다~

<img width="377" alt="image" src="https://github.com/zockq1/piggymy-frontend-admin/assets/13115713/2f7b1757-f3f4-46f1-9c61-7ab9799f0da0">

kangminhyuk1111 commented:
네 피드백 작업해서 다시올리겠습니다~

kangminhyuk1111 committed:
Merge branch 'zockq1:main' into feature/header

kangminhyuk1111 committed:
FEEDBACK: 피드백 반영

1. component 제거
2. console 제거
3. header ui 통합
4. layout으로 폴더 변경
5. interface로 타입 정의
6. fork 상태에서는 dev 브랜치를 받아오는게 아니라 main 브랜치를 받아오도록 되어있어서 일단 커밋하고 clone으로 작업방식을 변경하고 해야될 것 같습니다.
7. sidebar에 리스트를 추후에 정리해서 받아 올 필요가 있는 것 같습니다 다음에 회의때 한번 말해보는게 좋을 것 같습니다. 일단 unknown으로 올리겠습니다.

+ useeffect 내부에 console 찍은 이유는 git에 올릴때 lint때문에 사용하지 않는 변수는 올라가지 않게 되어있어서 그렇습니다. 일단 console로 찍어놓은거 다음에 데이터 받아올때 지우겠습니다.

zockq1 commented on src/share/ui/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,44 @@
+
+import { contentList } from '@/share/layout/header/headerList';
+
+interface SidebarProps {
+  sidebarList: unknown[];
+}
+
+export default function Sidebar({ sidebarList }: SidebarProps) {
+  const router = usePathname();
+  const activeSegments = router.split('/')[3];
+  const linkQuery: string[] = ['1', '2', '3', '4', '5'];
+
+  useEffect(() => {
+    console.log(sidebarList);
+  }, []);
+
```
1. 해당 구현 방법은 라우팅 경로에 대한 의존도가 높아보입니다.
```js
interface SidebarProps {
  menuItemList:{title: string; route: string; isCurrent: boolean}[];
}
```
  컴포넌트에서 훅을 사용해서 현재 라우팅에 따라 결과를 바꾸는 것보다는
  예를 들어 위 코드처럼 라우팅에 대한 정보(메뉴 목록, 현재 선택된 메뉴)를 props로 받은 뒤 출력하는게 좋아 보입니다

2. props에서 sidebar는 전체 컴포넌트를 뜻하는데 개별 메뉴에 대한 목록은 menuItemList, sidebarItemList등 별개의 이름을 쓰는게 좋아 보입니다 

3. 만양 위와 같은 방법으로 구현한다면 use client는 제거해도 좋을 것 같습니다

zockq1 replied:
```js

export const vocaQuizList: ListType[] = [
  { title: '용어카드 관리', route: '/test' },
  { title: '퀴즈카드 관리', route: '/test' },
];

interface SidebarProps {
  menuItemList:ListType[];
  currnetItemTitle: string
}
```

아니면 기존의ListType을 이용하고 추가로 currnetItemTitle을 넘겨주는것도 좋아보입니다

zockq1 commented on src/share/layout/header/Header.tsx:
```diff
@@ -0,0 +1,87 @@
+
+  useEffect(() => {
+    console.log(user);
+  });
```
use.userName값을 아래 관리자명(77줄)에 넣으면 Lint조건을 useEffect 없이 맞출 수 있을 것 같습니다

zockq1 commented on src/share/layout/header/Header.tsx:
```diff
@@ -0,0 +1,87 @@
+
+  const allDisplay = (state: boolean) => {
+    setDropdownDisplayState(!state);
+  };
```
1. 함수명과 내부 로직이 매칭되지 않습니다.
toggleDropdown처럼 명확한 의미를 가지게 하는게 좋을 것 같습니다.
2. 드랍다운 메뉴 외부를 클릭했을 때 닫히는 로직이 있으면 좋을 것 같습니다.

zockq1 commented on src/share/layout/header/Header.tsx:
```diff
@@ -0,0 +1,87 @@
+
+export default function Header({ user }: HeaderPropsType) {
+  const [dropdownDisplayState, setDropdownDisplayState] =
+    useState<boolean>(false);
```
2. [isDropdownOpen, setIsDropdownOpen] 처럼 더 명확한 변수, 함수명을 사용하는게 좋을 것 같습니다.

zockq1 commented on src/share/layout/header/Header.tsx:
```diff
@@ -0,0 +1,87 @@
+        className={`absolute top-20 ml-32 flex h-auto w-10/12 justify-center bg-white p-4 pt-2 ${dropdownDisplayState ? 'block' : 'hidden'}`}
+      >
+        <HeaderDetail headerList={[]} />
+      </div>
```
1. 드랍다운 관련 모든 로직(useState, 토글 함수 등)을 HeaderDetail 컴포넌트 내부로 옮기는게 좋아 보입니다.
현재는 드롭다운 메뉴를 여닫을 때마다 헤더 전체가 불필요하게 리렌더링 됩니다
```js
function HeaderDetail() {
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const toggleDropdown = () => {
    setIsDropdownOpen(!isDropdownOpen);
  };
  return <><button>전체보기</button>{isDropdownOpen && <>...</>}</>
}
```

zockq1 commented on src/share/layout/header/HeaderDetail.tsx:
```diff
@@ -0,0 +1,85 @@
+  vocaQuizList,
+} from '@/share/layout/header/headerList';
+
+export default function HeaderDetail({
+  headerList,
+}: {
+  headerList: unknown[];
+}) {
+  useEffect(() => {
+    console.log(headerList);
+  }, []);
```
headerList가 무슨 역할을 하는건지 모르겠습니다.

zockq1 commented on src/share/layout/header/HeaderDetail.tsx:
```diff
@@ -0,0 +1,85 @@
+
+  return (
+    <div className="flex h-full w-10/12 flex-row justify-center bg-white text-sm">
+      <ul className="h-full w-1/6">
+        <li className="bottom-1 mb-1 border-solid border-black text-center text-lg font-bold">
+          홈
+        </li>
+        {contentList.map((str, index) => (
+          <li key={index}>
+            <Link href={str.route}>{str.title}</Link>
+          </li>
+        ))}
+      </ul>
+      <ul className="h-full w-1/6">
+        <li className="bottom-1 mb-1 border-solid border-black text-center text-lg font-bold">
+          콘텐츠
+        </li>
+        {vocaQuizList.map((str, index) => (
+          <li key={index}>
+            <Link href={str.route}>{str.title}</Link>
+          </li>
+        ))}
+      </ul>
+      <ul className="h-full w-1/6">
+        <li className="bottom-1 mb-1 border-solid border-black text-center text-lg font-bold">
+          용어/퀴즈
+        </li>
+        {settingList.map((str, index) => (
+          <li key={index}>
+            <Link href={str.route}>{str.title}</Link>
+          </li>
+        ))}
+      </ul>
```
1. 피그마에서는 드롭다운 메뉴에 홈이 없습니다.
2. 홈으로 인해 변수가 1개씩 밀렸습니다(콘텐츠 -> vocaQuizList, 용어/퀴즈 -> settingList)

kangminhyuk1111 committed:
FEEDBACK: 피드백 반영 - 2

피드백 반영해서 다시 커밋 올립니다.
첫번째 피드백 주신 부분은 구현을 고민해봐야 할 것 같아서 일단 다른부분 먼저 작업 올립니다.
header에서 불필요한 랜더링이 일어나는 것 체크 해서 headerdetail에서 처리했습니다. 평소에 신경 안쓰고 작업했었는데 도움되는 코드리뷰였습니다 감사합니다.

+ 추가로 타입을 한번에 어떻게 관리하면 좋을지 궁금하네요. 파일을 세분화 할때마다 일일이 명시적으로 했었는데 type파일에 분리해놓고 사용하면 좋을 것 같다는 생각이 듭니다.

zockq1 commented on src/share/layout/header/HeaderDetail.tsx:
```diff
@@ -0,0 +1,57 @@
+  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
+  const dropdownRef = useRef<HTMLDivElement>(null);
+
+  const toggleDropdown = () => {
+    setIsDropdownOpen(!isDropdownOpen);
+  };
+
+  const handleClickOutside = (event: MouseEvent) => {
+    if (
+      dropdownRef.current &&
+      !dropdownRef.current.contains(event.target as Node)
+    ) {
+      setIsDropdownOpen(false);
+    }
+  };
+
+  useEffect(() => {
+    document.addEventListener('mousedown', handleClickOutside);
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, []);
```
드롭다운 열린 상태에서 전체보기 버튼 누르면 닫혀야 하는데 다시 열립니다.

드롭다운 열린 상태에서 전체보기 버튼 클릭 -> 이벤트 핸들러 작동으로 드롭다운 닫힘 -> button onClick 이벤트 작동으로 다시 열림

zockq1 commented on src/share/layout/header/HeaderDetail.tsx:
```diff
@@ -0,0 +1,57 @@
+    };
+  }, []);
+
+  useEffect(() => {
+    console.log(isDropdownOpen);
+  }, [isDropdownOpen]);
```
콘솔로그 지우고 커밋 부탁드립니다

zockq1 commented:
1. 전체적으로 피그마와 스타일이 다른 부분이 많습니다.
2. 사이드바는 share/layout, 로그인 컴포넌트는 (beforeLogin)/login/_component 디렉토리 만드셔서 거기에 넣어주시면 좋을 것 같습니다

kangminhyuk1111 committed:
FEEDBACK: 피드백 반영 - 로컬 push

kangminhyuk1111 committed:
Fix: 피드백 반영

사이드바 피드백 반영 했습니다.
동적으로 id 받아서 라우팅 되도록 구현했습니다.

사이드 바 폴더 위치 이동되었습니다.

type/testType.ts 폴더에 타입 정리 해놓아도 될까요?

kangminhyuk1111 committed:
Fix: 피드백 반영 - 2

+ 헤더 드롭다운 마우스 클릭 이벤트가 원활하게 작동이 안되는데 원인을 좀 찾아봐야 할 것 같아서 나중에 다시 올리겠습니다.

kangminhyuk1111 committed:
Fix: 피드백 반영 - 2

+ 헤더 드롭다운 마우스 클릭 이벤트가 원활하게 작동이 안되는데 원인을 좀 찾아봐야 할 것 같아서 나중에 다시 올리겠습니다.

kangminhyuk1111 committed:
Feat: dropdown

dropdown 작업중
기능은 완료 css만 추가하면 됩니다.
기능 관련 수정사항 있으면 알려주세요~

kangminhyuk1111 committed:
local push - 리뷰xxxxxx

zockq1 commented on src/share/layout/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,54 @@
+              );
+            })
+          : null}
+        <li className="pb-2 pl-4 text-lg font-bold">Route에 따라 변경</li>
+        {contentList.map((item, index) => {
+          const isActive = activeSegments == item.query;
+          return (
+            <li
+              key={index}
+              className={`p-2 ${isActive ? 'bg-primary' : 'bg-gray-3'}`}
+            >
+              <Link href={`/test/route/${item.query}`} className="pl-2">
+                {item.title}
+              </Link>
+            </li>
+          );
+        })}
+        <li></li>
```
해당 코드는 콘텐츠만 축력하는 테스팅용 코드라 지워야 할 것 같습니다.

zockq1 commented on src/share/layout/header/headerListData.ts:
```diff
@@ -0,0 +1,37 @@
+type ListType = {
+  title: string;
+  route: string;
+  query: string;
```
query 없이 route에 전체 주소 작성하시면 될 것 같습니다

zockq1 commented on src/share/layout/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,54 @@
+}
+
+export default function Sidebar({ sidebarList }: SidebarProps) {
+  const route = useParams();
+  const activeSegments = route.testId;
```
현재 경로를 props로 내려주지 않고 routing 경로에 의존하게 되면 스토리북 같은 테스팅 라이브러리 사용 시에 별도의 라우팅 설정을 해야 할 것 같은데, 해당 부분 고려해서 스토리 작성 부탁드립니다

zockq1 replied:
이건 다이나믹 라우팅을 사용하신 것 같은데 실제 구현 시에는 사이드 메뉴마다 전부 다른 요소를 가지기에
실제로는 사용 불가능 할 것 같습니다.

제가 위에 작성한것처럼 주소 전체를 비교해야 할 것 같습니다

zockq1 commented on src/share/layout/sidebar/Sidebar.tsx:
그리고 전체적으로 피그마 디자인이랑 너무 다른데 수정 부탁드립니다
(보더, 배경색, 폰트 글자색 등)

zockq1 commented on src/share/layout/header/HeaderDetail.tsx:
전체적으로 피그마 디자인 처럼 수정 부탁드립니다

zockq1 commented on src/share/layout/header/HeaderList.tsx:
```diff
@@ -0,0 +1,26 @@
+        {title}
+      </li>
+      {items.map((item: HeaderItem, index: number) => (
+        <li key={index}>
```
key에는 index보다는 item.title 등 고유값을 넣는 게 좋습니다.
자세하게는 react reconciliation 검색해 보시면 좋습니다

zockq1 commented on src/share/layout/sidebar/Sidebar.tsx:
```diff
@@ -0,0 +1,54 @@
+        {/* //TODO 테스트 리스트 추후변경*/}
+        {sidebarList
+          ? sidebarList.map((item: ListType, index: number) => {
+              const isActive = activeSegments == item.query;
+              return (
+                <li
+                  key={index}
+                  className={`p-2 ${isActive ? 'bg-primary' : 'bg-gray-3'}`}
+                >
+                  <Link href={`/test/route/${item.query}`} className="pl-2">
+                    {item.title}
+                  </Link>
+                </li>
+              );
```
query 제거하고 route에 전체 주소 보내서
```js
const route = usePathname()
...
const isActive = route.includes(item.route)
...
<Link href={route} className="pl-2">
```

로 현재 주소를 비교하고, route로 바로 이동하는게 좋을 것 같습니다

kangminhyuk1111 committed:
Feat: 로그인, 드롭다운 작업
Fix: PR 리뷰 반영

item.route로 통일해서 라우팅 구현 했습니다.
피그마랑 디자인 다른 부분은 수정중입니다.
로그인 배율 확대 축소시 벌어지는 부분 있습니다.

+ 로그인 아이콘이 깨져서 디자이너님 한테 따로 물어봐야 할 것 같습니다.

+ 드롭다운 css 미완성입니다.

+ header에 하드코딩 안하고 border 속성이랑 큰 카테고리를 감싸는 법을 좀 생각해 봐야 할 것 같습니다. 추후 작업하겠습니다.

zockq1 commented on src/share/layout/header/Header.tsx:
```diff
@@ -0,0 +1,85 @@
+      </div>
+      <div className="flex h-full w-8/12 flex-row">
+        <ul className="flex w-full flex-row items-center justify-around">
+          <li>
+            <Link
+              className="text-sm font-bold hover:text-blue-5"
+              href="/public"
+            >
+              홈
+            </Link>
+          </li>
+          <li>
+            <Link
+              className="text-sm font-bold hover:text-blue-5"
+              href="/public"
+            >
+              콘텐츠
+            </Link>
+          </li>
+          <li>
+            <Link
+              className="text-sm font-bold hover:text-blue-5"
+              href="/public"
+            >
+              용어/퀴즈
+            </Link>
+          </li>
```
이부분 반복적으로 나오고, 라우팅 연결도 안되어 있는데 아래처럼 구현하는게 좋을 것 같습니다.
```js
//
export const headerRouteList: ListType[] = [
  { title: '홈', route: '/admin' },
  { title: '콘텐츠', route: '/admin/content/greetingMessage' },
  { title: '용어/퀴즈', route: '/admin/content/themeCard' },
   ...
];
...
{headerRouteList.map((headerRoute) => {
    return <li key={headerRoute.title}>
            <Link
              className="text-sm font-bold hover:text-blue-5"
              href={headerRoute.route}
            >
             {headerRoute.title}
            </Link>
          </li>
  })
}
```

zockq1 commented on src/share/layout/header/headerListData.ts:
해당 파일은 header에서만 쓰이는게 아닌 것 같은데

이름은 share/route/route.ts 로 해서 폴더를 분리하는게 어떨까요?

zockq1 commented on src/share/layout/header/headerListData.ts:
```diff
@@ -0,0 +1,36 @@
+type ListType = {
```
ListType 보다는 route에 관한 타입이니까 RouteModel, Route 등의 이름이 좋을 것 같습니다.

그리고 네이밍 할때 보통 함수 뒤에 Function 붙이지 않고 변수 뒤에 Variable 붙이지 않는 것처럼 타입 뒤에 Type을 붙이는 것은 조금 이상해 보입니다. 개인적으로는 간결하게 Route 추천 드립니다.

zockq1 commented on src/share/layout/header/HeaderList.tsx:
```diff
@@ -0,0 +1,31 @@
+interface HeaderItem {
+  route: string;
+  title: string;
+}
```
이미 ListType 정의하셨으니 해당 타입을 export해서 사용하는게 좋을 것 같습니다

zockq1 commented on src/share/layout/header/headerListData.ts:
```diff
@@ -0,0 +1,36 @@
+type ListType = {
+  title: string;
+  route: string;
+};
```
해당 부분을 routeType으로 분리해서 type 폴더에 놓는게 좋을 것 같습니다.

kangminhyuk1111 committed:
Merge remote-tracking branch 'upstream/develop' into feature/header

kangminhyuk1111 committed:
Feat: storybook 작업

스토리북 관련 작업 해놓았습니다.

+ lint 설정 때문인지 모르겠는데 fix 하면 commit에 같이 들어갑니다.

kangminhyuk1111 committed:
Fix: 피드백 내용 작업

kangminhyuk1111 committed:
local push
